File: justjavastaff/src/main/java/com/marknkamau/justjavastaff/data/network/OrdersRepositoryImpl.java
package com.marknkamau.justjavastaff.data.network;

import android.support.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.FirebaseFirestoreSettings;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.QuerySnapshot;
import com.marknkamau.justjavastaff.models.Order;
import com.marknkamau.justjavastaff.models.OrderItem;
import com.marknkamau.justjavastaff.models.OrderStatus;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

import timber.log.Timber;

/**
 * Created by Mark Njung'e.
 * mark.kamau@outlook.com
 * https://github.com/MarkNjunge
 */

public class OrdersRepositoryImpl implements OrdersRepository {
    private FirebaseFirestore firestore;

    public OrdersRepositoryImpl() {
        firestore = FirebaseFirestore.getInstance();
        FirebaseFirestoreSettings settings = new FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build();

        firestore.setFirestoreSettings(settings);
    }

    @Override
    public void getOrders(final OrdersListener listener) {
        ListenerRegistration registration = firestore.collection("orders")
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
                    @Override
                    public void onEvent(QuerySnapshot querySnapshot, FirebaseFirestoreException e) {
                        if (e != null) {
                            Timber.e(e);
                            listener.onError(e.getMessage());
                        } else {
                            List<Order> orders = new ArrayList<>();
                            List<DocumentSnapshot> documentSnapshots = querySnapshot.getDocuments();

                            for (DocumentSnapshot document : documentSnapshots) {
                                Map<String, Object> data = document.getData();
                                Order order = new Order(
                                        (String) data.get("orderId"),
                                        (String) data.get("customerName"),
                                        (String) data.get("customerPhone"),
                                        (String) data.get("deliveryAddress"),
                                        (String) data.get("additionalComments"),
                                        (String) data.get("status"),
                                        (Date) data.get("timestamp"),
                                        (int) (long) data.get("totalPrice"),
                                        (int) (long) data.get("itemsCount")
                                );

                                orders.add(order);
                            }

                            listener.onSuccess(orders);
                        }
                    }
                });
    }

    @Override
    public void getOrderItems(String orderId, final OrderItemsListener listener) {
        firestore.collection("orderItems").whereEqualTo("orderId", orderId)
                .get()
                .addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot querySnapshot) {
                        List<OrderItem> items = new ArrayList<>();
                        List<DocumentSnapshot> documents = querySnapshot.getDocuments();

                        for (DocumentSnapshot document : documents) {
                            Map<String, Object> data = document.getData();
                            OrderItem orderItem = new OrderItem(
                                    data.get("itemName").toString(),
                                    (int) (long) data.get("itemQty"),
                                    (Boolean) data.get("itemCinnamon"),
                                    (Boolean) data.get("itemChoc"),
                                    (Boolean) data.get("itemMarshmallow"),
                                    (int) (long) data.get("itemPrice")
                            );
                            items.add(orderItem);
                        }

                        listener.onSuccess(items);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Timber.e(e);
                        listener.onError(e.getMessage());
                    }
                });
    }

    @Override
    public void updateOderStatus(String orderId, OrderStatus status, final BasicListener listener) {
        firestore.collection("orders").document(orderId)
                .update("status", status.name())
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        listener.onSuccess();
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        listener.onError(e.getMessage());
                    }
                });
    }

}

File: justjavastaff/src/main/java/com/marknkamau/justjavastaff/data/network/OrdersRepositoryImpl.kt
package com.marknkamau.justjavastaff.data.network

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreSettings
import com.marknkamau.justjavastaff.models.Order
import com.marknkamau.justjavastaff.models.OrderItem
import com.marknkamau.justjavastaff.models.OrderStatus

import java.util.ArrayList
import java.util.Date

import timber.log.Timber

/**
 * Created by Mark Njung'e.
 * mark.kamau@outlook.com
 * https://github.com/MarkNjunge
 */

class OrdersRepositoryImpl : OrdersRepository {
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()

    init {
        val settings = FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .build()

        firestore.firestoreSettings = settings
    }

    override fun getOrders(listener: OrdersRepository.OrdersListener) {
        val registration = firestore.collection("orders")
                .addSnapshotListener { querySnapshot, e ->
                    if (e != null) {
                        Timber.e(e)
                        listener.onError(e.message ?: "Error getting orders")
                    } else {
                        val orders = ArrayList<Order>()
                        val documentSnapshots = querySnapshot?.documents

                        documentSnapshots?.let {

                            for (document in documentSnapshots) {
                                val data = document.data
                                data?.let {

                                    val order = Order(
                                            data["orderId"] as String,
                                            data["customerName"] as String,
                                            data["customerPhone"] as String,
                                            data["deliveryAddress"] as String,
                                            data["additionalComments"] as String,
                                            data["status"] as String,
                                            data["timestampNow"] as Date,
                                            (data["totalPrice"] as Long).toInt(),
                                            (data["itemsCount"] as Long).toInt()
                                    )
                                    orders.add(order)
                                }
                            }

                            listener.onSuccess(orders)
                        }
                    }
                }
    }

    override fun getOrderItems(orderId: String, listener: OrdersRepository.OrderItemsListener) {
        firestore.collection("orderItems").whereEqualTo("orderId", orderId)
                .get()
                .addOnSuccessListener { querySnapshot ->
                    val items = ArrayList<OrderItem>()
                    val documents = querySnapshot.documents

                    for (document in documents) {
                        val data = document.data
                        data?.let {
                            val orderItem = OrderItem(
                                    data["itemName"].toString(),
                                    (data["itemQty"] as Long).toInt(),
                                    data["itemCinnamon"] as Boolean,
                                    data["itemChoc"] as Boolean,
                                    data["itemMarshmallow"] as Boolean,
                                    (data["itemPrice"] as Long).toInt()
                            )
                            items.add(orderItem)
                        }
                    }

                    listener.onSuccess(items)
                }
                .addOnFailureListener { e ->
                    Timber.e(e)
                    listener.onError(e.message ?: "Error getting order items")
                }
    }

    override fun updateOderStatus(orderId: String, status: OrderStatus, listener: OrdersRepository.BasicListener) {
        firestore.collection("orders").document(orderId)
                .update("status", status.name)
                .addOnSuccessListener { listener.onSuccess() }
                .addOnFailureListener { e ->
                    listener.onError(e.message ?: "Error updating order status")
                }
    }

}
